### DISCUSSION TITLE
Typing improvements in Middleware.

### DISCUSSION THREAD

**Discussion by pawelrubin** [encode/starlette#2380] [Q&A]
Hi everyone!

Using [`Middleware`](https://www.starlette.io/middleware/) or `add_middleware` hides the signature of the Middleware class being passed.

For example, the following code:

```python
from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.middleware.sessions import SessionMiddleware
from starlette.responses import JSONResponse
from starlette.routing import Route
from starlette.requests import Request


async def homepage(request: Request) -> JSONResponse:
    return JSONResponse({"hello": "world"})


app = Starlette(
    debug=True,
    routes=[
        Route("/", homepage),
    ],
    middleware=[
        Middleware(SessionMiddleware),
    ],
)
```

raises a Runtime exception:
```shell
TypeError: SessionMiddleware.__init__() missing 1 required positional argument: 'secret_key'
```

but no typing errors.

---

This can be improved by leveraging `ParamSpec` introduced in Python 3.10. 
Here's a reference implementation in my fork: https://github.com/pawelrubin/starlette/commit/0f4c8c5d93359e06148a4a436c113ee68bba82e3

I'd love to hear the Maintainers' and Community's feedback regarding this issue.

**Comment by adriangb** [comment]
Please submit a PR and we can look at it 😄

**Reply by pawelrubin** [reply]
Sure :) https://github.com/encode/starlette/pull/2381
